Add hdf5 julia support to datascience stack

mkolovic opened the pull request:
Fixes https://github.com/jupyter/docker-stacks/issues/165
Replaces https://github.com/jupyter/docker-stacks/pull/157

* add `.juliarc.jl` allowing Julia to locate hdf5 libraries installed by `conda` via h5py
* install HDF5 Julia package 
* update README 

This is meant to fix issue #165 .

jakirkham commented:
Thanks @mkolovic. Let's see how the tests go. Did you try this out locally?

mkolovic commented:
I did build an image locally using this branch, seems to be working well.

mkolovic commented:
I may have spoken too soon. When I build the Docker image using this branch the `.juliarc.jl` is present correctly, but the HDF5 install fails but lets the build proceed. It will let you build HDF5 immediately upon launching a notebook. So the rc file is good, it just isn't getting recognized at build time. I suspect because the `echo` to  `.juliarc.jl` and the `Pkg.add()` are in the same `RUN` statement.

jakirkham commented:
> When I build the Docker image using this branch the `.juliarc.jl` is present correctly, but the HDF5 install fails but lets the build proceed.

Good. So, we are on the right path here.

> I suspect because the `echo` to  `.juliarc.jl` and the `Pkg.add()` are in the same `RUN` statement.

Shouldn't cause a problem. Though if you do think it is the issue, have you tried breaking them out into two `RUN` steps?

Maybe Julia needs to be told to refresh what it knows about the environment and what is installed somehow.

mkolovic commented:
The `RUN` split had no effect and you're right, it shouldn't. Looks like julia doesn't load `.juliarc.jl` when using the `-e` flag. I couldn't find definitive documentation on that. `-F` forces loading the config file.

jakirkham commented:
>  `-F` forces loading the config file.

Seems like that was the missing part. Just to confirm, this gives you a working copy of the HDF5 package for Julia out of the box?

Also, would you mind squashing the docker commits and then force pushing?

mkolovic commented:
Launching a fresh instance of the Docker image and opening a Julia notebook:
```
using HDF5
In [4]:

A = reshape(1:120, 15, 8)
Out[4]:
15x8 Array{Int64,2}:
  1  16  31  46  61  76   91  106
  2  17  32  47  62  77   92  107
  3  18  33  48  63  78   93  108
  4  19  34  49  64  79   94  109
  5  20  35  50  65  80   95  110
  6  21  36  51  66  81   96  111
  7  22  37  52  67  82   97  112
  8  23  38  53  68  83   98  113
  9  24  39  54  69  84   99  114
 10  25  40  55  70  85  100  115
 11  26  41  56  71  86  101  116
 12  27  42  57  72  87  102  117
 13  28  43  58  73  88  103  118
 14  29  44  59  74  89  104  119
 15  30  45  60  75  90  105  120
In [5]:

h5write("/home/jovyan/work/test.h5", "mygroup2/A", A)
In [7]:

data = h5read("/home/jovyan/work/test.h5", "mygroup2/A", (2:3:15, 3:5))
Out[7]:
5x3 Array{Int64,2}:
 32  47  62
 35  50  65
 38  53  68
 41  56  71
 44  59  74
```

So it seems the basic function is there at least. 

> Also, would you mind squashing the docker commits and then force pushing?

I am not entirely sure what you mean. Squashing the images as in reducing their sizes? Pushing to where?

Thanks.

rgbkrk commented:
@jakirkham did you meant squishing the git commits?

jakirkham commented:
Yep, that's what I meant. Sorry for being unclear.

rgbkrk commented:
It's all good. :smile:

mkolovic commented:
Ah ok, I think I squashed successfully using `git rebase -i HEAD~5` and then `git push -f`. While I was figuring it out I ended up merging the `upstream/master`, that is the `jupyter/docker-stacks/master`, into the 'add-hdf-julia` branch on my fork. I thought I should get the latest updates into that branch before pushing again, but now I see it wasn't necessary. I squashed that commit as well and suppresed the commit message. I hope there is no negative side effect.

If I may ask, is there a method to keep a branch on a fork updated with the master branch of where the project was forked from? I suppose for long standing branches on the fork this might be of interest.


Thank you for the help and support, it is much appreciated.

jakirkham commented:
AFAICT, it looks like you have done the right thing. Thanks @mkolovic.

Unless there are merge conflicts, it shouldn't be necessary to pull in upstream changes. Travis creates a merge commit with `master` or whatever branch you are PR-ing against automatically. GitHub will handle merging correctly as long as there are no conflicts.

There have been suggestions to GitHub to add a button to update branches on forks with the original. This still hasn't happened, but it might one day. Everyone is entitled to git workflow that works well for them. Normally, I try never to commit directly to `master` and always use a branch. That way, I can have `master` track the upstream repo and push the changes into my fork as they become available. All my branches are then various bug fixes or features that I can PR when ready. If you have more questions, I can try to provide an example.

Thanks for being so patience and persistent, @mkolovic. I am good with this change once CI passes.

mkolovic commented:
That makes perfect sense, thanks again.

jakirkham commented:
Sure, glad to help. This change has now been release to Docker Hub.

parente commented:
@mkolovic Thanks again for the PR! :balloon:
